Make the test cluster works through Shelley and Allegra / cardano-node-1.24.0 bump

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

This seems to work node-wise (although I am using the latest revision from the node, which should be soon tagged as 1.24.0). A few notes:

  - We need to adjust the code for constructing transactions (in particular
    'mkUnsignedTx') since they are serialized in slightly different
    ways. For immediate testing purposes, we may "brute-force" this and
    simply use the Allegra era format all the time. But in practice, we
    may want to release a version of the wallet that can do both and
    cross the hard fork nicely. Maybe there's something we can use from the cardano-api 
    to make this easier :thinking: ?

  - The launcher code needs some cleanup so that all the commands drop
    the now-deprecated 'shelley' prefix and use top-level commands
    instead.

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
Make the test cluster works through Shelley and Allegra.

  This seems to work node-wise (although I am using the latest revision
  from the node, which should be soon tagged as 1.24.0). A few notes:

  - We need to adjust the code for constructing transactions (in particular
    'mkUnsignedTx') since they are serialized in slightly different
    ways. For immediate testing purposes, we may "brute-force" this and
    simply use the Allegra era format all the time. But in practice, we
    may want to release a version of the wallet that can do both and
    cross the hard fork nicely.

  - The launcher code needs some cleanup so that all the commands drop
    the now-deprecated 'shelley' prefix and use top-level commands
    instead.

Co-authored-by: @Anviking <johannes.lund@iohk.io>
Co-authored-by: @KtorZ <matthias.benkort@iohk.io>
Co-authored-by: @rvl <rodney.lorrimar@iohk.io>

Anviking commented on lib/shelley/test/data/cardano-node-shelley/node.config:
```diff
@@ -26,13 +26,13 @@ PBftSignatureThreshold: 1
-TestShelleyHardForkAtVersion: 1
-
+TestShelleyHardForkAtEpoch: 1
+TestAllegraHardForkAtEpoch: 2
```
The original intention of the `withCluster` was to allow the caller to run code on each era of the chain. Looking now, perhaps the existing doc comment doesn't make it clear, but:

could we still add a note about the fork(s) being triggered automatically in the documentation for `withCluster`?

rvl replied:
Yes, if we need the wallet to work in two eras then we ought to run some tests in both eras.
But perhaps while the wallet is effectively allegra-only it can be simpler and use the hard fork schedule.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -1758,8 +1759,8 @@ instance ToText ClusterLog where
-        MsgWaitingForFork ->
-            "Transitioning from Byron to Shelley... Please wait 20s..."
+        MsgWaitingForFork fork ->
+            "Transitioning from " <> fork <> " ... Please wait 20s..."
```
👍

KtorZ committed:
Update dependencies to cardano-1.24.0

Co-authored-by: @Anviking <johannes.lund@iohk.io>
Co-authored-by: @KtorZ <matthias.benkort@iohk.io>
Co-authored-by: @rvl <rodney.lorrimar@iohk.io>

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -747,14 +778,47 @@ fromShelleyTx (SL.Tx bod@(SL.TxBody ins outs certs wdrls _ _ _ _) _ mmd) =
+    SL.Tx bod@(SL.TxBody ins outs certs wdrls _ _ _ _) _ mmd = tx
+
+fromAllegraTx
+    :: ( SL.ShelleyBased era
+       , SL.Core.TxBody era ~ MA.TxBody era
+       , SL.Core.Value era ~ SL.Coin
+       , SL.Core.Metadata era ~ MA.Metadata era
+       , Ord (SL.Core.Script era)
+       )
```
Perhaps there is a single AllegraEra type that satifies all of these constraints?

rvl reviewed (approved):
Looks good to me, but sometimes it fails to register the stake pools. I am trying to see why that is.

KtorZ committed:
make transactions 'pure Allegra' transactions

  There seems to be an issue with the node when in Allegra era which does not process Shelley transactions despite accepting them. The issue is under investigation from the node's team but in the meantime, this commit may helps us testing that the wallet does work corretly in Allegra. This is not a commit we likely want to keep and ship in the next version because it'll prevent users from sending any transaction until the hard-fork happens.

KtorZ committed:
log server errors as critical log messages when an exception is raised

  It took us long enough to get to identify the source of the failure because the wallet is silently raising an exception and terminating nodes of the cluster, but there are some ghost threads which remain hanging and prevent other bracket functions to fully move forward and stop.
  At least now, an error message is printed to help debugging the root cause of the issue.

KtorZ committed:
fix 'toCardanoTxOut' to use alternative in case it fails parsing a Byron address

  'deserialiseFromRawBytes AsShelleyAddress' used to work transparently
  for both Shelley and Byron (so-called Bootstrap) addresses. This is no
  longer the case so we need to alternatively try 'AsShelleyAddress' and
  'AsByronAddress' to deserialise any addresses. See also:

  ```
  commit cebfd99763cbdefd2856def75e27a0300b79a250
  Author: Duncan Coutts <duncan@well-typed.com>
  Date:   Fri Nov 20 01:14:21 2020 +0000
  ```

  on `input-output-hk/cardano-node`

KtorZ committed:
make the era dynamic within the network layer + 'getCurrentEra'

  It is a bit painful and _might be simplifiable_. The annoying point
  comes with the fact that as soon as we're making one query, we are
  fixing the type of the result within the scope of the function we are
  in, yet, we typically want to work 'forall era' and providing type
  signatures for the nested query types is quite cumbersome. What I
  found simpler to read and comprehend was simply to pattern match on
  the era type and run execute queries + transformation within the same
  branch of the pattern match, coming at the cost of a few duplications.

  I also took the opportunity to add a `getCurrentEra' to the network
  layer, which should help us to get the current era in other places
  (e.g. the transaction layer) when needed. That era is updated after
  every tip update.

rvl committed:
test cluster: Try to get wallet file logging on by default

Anviking committed:
Switch to QueryIfCurrentAllegra

KtorZ committed:
make the transaction layer able to handle multiple eras.

rvl committed:
Fix compilation of tests

I wanted the transaction specs to be parameterised by era, but
couldn't work out how.

rvl committed:
Parameterise prop_decodeSignedShelleyTxRoundtrip by shelley era FWIW

rvl committed:
Add currentNodeEra to WalletSpec dummyNetworkLayer

rvl committed:
stack.yaml: Update to latest cardano-node 1.24.1 release branch

rvl committed:
stack.yaml: Update to latest cardano-node 1.24.1

rvl committed:
Regenerate cabal.project and cabal.project.freeze

rvl committed:
nix: Add script for invoking stack2cabal

rvl committed:
Fix cabal configure check in buildkite

Anviking committed:
Try lower parallelism from -j 8 to -j 3 in CI

And see if this makes tests pass.
